File: Haesoo/Week16/BOJ_다오의데이트.java
import java.util.*;
import java.io.*;

public class 다오 {
    public static int H, W, N;
    public static int[] dy = {-1, 0, 1, 0};
    public static int[] dx = {0, 1, 0, -1};
    public static char[][] m = new char[21][21];
    public static char[] command = {'W', 'D', 'S', 'A'};
    public static List <Integer> ans = new ArrayList<>();
    public static int len = 10000000;
    public static pair start;
    public static boolean flag = false;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        H = Integer.parseInt(st.nextToken());
        W = Integer.parseInt(st.nextToken());
        for (int i = 1; i <= H; i++) {
            //st = new StringTokenizer(br.readLine());
            String s = br.readLine();
            for (int j = 1; j <= W; j++) {
                m[i][j] = s.charAt(j - 1);// st.nextToken().charAt(0);
                if (m[i][j] == 'D') {
                    start = new pair(i, j);
                    m[i][j] = '.';
                }
            }
        }
        N = Integer.parseInt(br.readLine());
        pair[] v = new pair[N];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            char c1 = st.nextToken().charAt(0);
            char c2 = st.nextToken().charAt(0);
            v[i] = new pair(convert(c1), convert(c2));
        }
        List <Integer> path = new ArrayList<>();

        dfs(v, path, start, 0);
        if (flag) return;
        if (len == 10000000) {
            System.out.print("N0");
            return;
        }

    }
    public static int convert (char c) {
        if (c == 'W') return 0;
        else if (c == 'D') return 1;
        else if (c == 'S') return 2;
        else return 3;
    }
    public static boolean range (int ny, int nx) {
        return ny >= 0 && nx >= 0 && ny <= H && nx <= W;
    }
    public static void dfs (pair[] v, List <Integer> path, pair now, int depth) {
        if (m[now.first][now.second] == 'Z') {
            if (path.size() < len) {
                len = path.size();
                ans = path;
                flag = true;
                System.out.println("YES");
                for (int i = 0; i < ans.size(); i++) {
                    System.out.print(command[ans.get(i)]);
                }
            }
            return;
        }
        if (depth == N) return;
        int ny = now.first + dy[v[depth].first];
        int nx = now.second + dx[v[depth].first];
        if (range(ny, nx) && m[ny][nx] != '@') {
            path.add(v[depth].first);
            dfs(v, path, new pair (ny, nx), depth + 1);
            path.remove(path.size() - 1);
        }

        ny = now.first + dy[v[depth].second];
        nx = now.second + dx[v[depth].second];
        if (range(ny, nx) && m[ny][nx] != '@') {
            path.add(v[depth].second);
            dfs(v, path, new pair (ny, nx), depth + 1);
            path.remove(path.size() - 1);
        }

    }
    public static class pair {
        int first;
        int second;
        pair (int f, int s) {
            this.first = f;
            this.second = s;
        }
    }
}

File: Haesoo/Week16/BOJ_다오의데이트.cpp
#include <bits/stdc++.h>
#include <iostream>
using namespace std;

const int MAX = 21;
int h, w, n, dir[4][2] = { {-1, 0}, {0, 1}, {1, 0}, {0, -1} }, len = INT_MAX;
char m[MAX][MAX], rev[4] = {'W', 'D', 'S', 'A'};
bool visited[MAX][MAX];
pair<int, int> s;
vector<int> ans;

int convert(char c) {
    switch (c) {
    case 'W':
        return 0;
    case 'D':
        return 1;
    case 'S':
        return 2;
    case 'A':
        return 3;
    }
}

void func(vector<pair<int, int>>& v, vector<int>& path, pair<int, int> now, int idx) {
    if (m[now.first][now.second] == 'Z') {
        if (path.size() < len) {
            len = path.size();
            ans = path;
        }

    return;
}

if (idx == v.size())
    return;

int nr = now.first + dir[v[idx].first][0];
int nc = now.second + dir[v[idx].first][1];

if (nr >= 1 && nr <= h && nc >= 1 && nc <= w && m[nr][nc] != '@') {
    path.push_back(v[idx].first);
    func(v, path, { nr, nc }, idx + 1);
    path.pop_back();
}

nr = now.first + dir[v[idx].second][0];
nc = now.second + dir[v[idx].second][1];

if (nr >= 1 && nr <= h && nc >= 1 && nc <= w && m[nr][nc] != '@') {
    path.push_back(v[idx].second);
    func(v, path, { nr, nc }, idx + 1);
    path.pop_back();
}
}

int main(void) {
    ios::sync_with_stdio(false);cin.tie(0);cout.tie(0);

cin >> h >> w;
for (int i = 1; i <= h; i++) {
    for (int j = 1; j <= w; j++) {
        char c;
        cin >> c;
        if (c == 'D') {
            s = { i, j };
            m[i][j] = '.';
            continue;
        }

        m[i][j] = c;
    }
}

cin >> n;
vector<pair<int, int>> v(n);
for (auto& i : v) {
    char c1, c2;
    cin >> c1 >> c2;
    i = { convert(c1), convert(c2) };
}

vector<int> path;
func(v, path, s, 0);
if (len == INT_MAX) {
    cout << "NO\n";
    return 0;
}

cout << "YES\n";
for (auto& i : ans)
    cout << rev[i];

cout << endl;
return 0;
}
